library/std/src/sys/unix/locks/futex.rs

```diff
@@ -0,0 +1,159 @@
+use crate::sync::atomic::{
+    AtomicI32,
+    Ordering::{Acquire, Relaxed, Release},
+};
+use crate::sys::futex::{futex_wait, futex_wake, futex_wake_all};
+use crate::time::Duration;
+
+pub type MovableMutex = Mutex;
+pub type MovableCondvar = Condvar;
+
+pub struct Mutex {
+    /// 0: unlocked
+    /// 1: locked, no other threads waiting
+    /// 2: locked, and other threads waiting (contended)
+    futex: AtomicI32,
```

Comment:
Can/should this be i8? Iirc, one of the benefits of WTFLock-style mutexes is that they have a small size_of, which allows embedding them into, eg, every JS object.

Reply:
Nope, futexes are always 32 bit on Linux. See https://man7.org/linux/man-pages/man2/futex.2.html

Reply:
It's possible by using a bitmask on the futex (shifting the bitmask so only the byte we're interested in is affected).

Reply:
> It's possible by using a bitmask on the futex (shifting the bitmask so only the byte we're interested in is affected).

Is there any guarantee that a 32 bit futex only has 8 effective bits so that it can be embedded in an i8?

Reply:
The futex `uaddr` is also required to be 4-byte aligned.

Reply:
> I think using a bitmask will still cause the full 32bit to be loaded by the kernel. If the futex is in the last byte of a page with the next page unmapped that would cause an EFAULT.

Aligning the address and shifting the mask so it only covers the required byte works, though (even if it is borderline UB).

Reply:
[futex2](https://www.kernel.org/doc/html/latest/userspace-api/futex2.html#types-of-futex) will add variable-sized futexes (i8 / i16 / i32 / i64) to linux, but currently we only have 32 bit ones.